From dee4cd0b0523c682d34f943237371ac42610cd5f Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Tue, 23 Jan 2018 18:49:46 +0000
Subject: [PATCH] Add support for a stateless configuration file

Allow udisks to look at /usr/share/defaults/udisks/udisks2.conf in
case the /etc/udisks/udisks2.conf file does not exist.
---
 src/udisksconfigmanager.c | 52 ++++++++++++++++++++++++++++-----------
 1 file changed, 37 insertions(+), 15 deletions(-)

diff --git a/src/udisksconfigmanager.c b/src/udisksconfigmanager.c
index 190852ee..e1084978 100644
--- a/src/udisksconfigmanager.c
+++ b/src/udisksconfigmanager.c
@@ -171,7 +171,9 @@ udisks_config_manager_constructed (GObject *object)
   UDisksConfigManager *manager = UDISKS_CONFIG_MANAGER (object);
   GError *error = NULL;
   GKeyFile *config_file;
-  gchar *conf_filename;
+  gchar *conf_filename = NULL;
+  gchar *user_conf_filename;
+  gchar *system_conf_filename;
   gchar *load_preference;
   gchar *encryption;
   gchar *module_i;
@@ -184,23 +186,42 @@ udisks_config_manager_constructed (GObject *object)
   g_key_file_set_list_separator (config_file, ',');
 
   /* Get modules and means of loading */
-  conf_filename = g_build_filename (G_DIR_SEPARATOR_S,
-                                    udisks_config_manager_get_uninstalled (manager) ?
-                                      BUILD_DIR :
-                                      PACKAGE_SYSCONF_DIR,
-                                    udisks_config_manager_get_uninstalled (manager) ?
-                                      "udisks" :
-                                      PROJECT_SYSCONF_DIR,
-                                    PACKAGE_NAME_UDISKS2 ".conf",
-                                    NULL);
-
-  udisks_debug ("Loading configuration file: %s", conf_filename);
+  user_conf_filename = g_build_filename (G_DIR_SEPARATOR_S,
+                                         udisks_config_manager_get_uninstalled (manager) ?
+                                           BUILD_DIR :
+                                           PACKAGE_SYSCONF_DIR,
+                                        udisks_config_manager_get_uninstalled (manager) ?
+                                          "udisks" :
+                                           PROJECT_SYSCONF_DIR,
+                                         PACKAGE_NAME_UDISKS2 ".conf",
+                                         NULL);
+
+  system_conf_filename = g_build_filename (G_DIR_SEPARATOR_S,
+                                           "/usr/share/defaults",
+                                           PROJECT_SYSCONF_DIR,
+                                           PACKAGE_NAME_UDISKS2 ".conf",
+                                           NULL);
+
+
+  udisks_debug ("Loading configuration file: %s", user_conf_filename);
 
   /* Load config */
   if (g_key_file_load_from_file (config_file,
-                                 conf_filename,
+                                 user_conf_filename,
                                  G_KEY_FILE_NONE,
                                  &error))
+    {
+      conf_filename = user_conf_filename;
+    }
+  else if (g_key_file_load_from_file (config_file,
+                                      system_conf_filename,
+                                      G_KEY_FILE_NONE,
+                                      &error))
+    {
+      conf_filename = system_conf_filename;
+    }
+
+  if (conf_filename)
     {
       modules = g_key_file_get_string_list (config_file,
                                             modules_group_name,
@@ -305,7 +326,7 @@ udisks_config_manager_constructed (GObject *object)
   else
     {
       g_clear_error (&error);
-      udisks_warning ("Can't load configuration file %s", conf_filename);
+      udisks_warning ("Can't load configuration file %s", system_conf_filename);
       manager->modules = NULL; /* NULL == '*' */
       manager->load_preference = UDISKS_MODULE_LOAD_ONDEMAND;
       manager->encryption = UDISKS_ENCRYPTION_DEFAULT;
@@ -313,7 +334,8 @@ udisks_config_manager_constructed (GObject *object)
 
 
   g_key_file_free (config_file);
-  g_free (conf_filename);
+  g_free (user_conf_filename);
+  g_free (system_conf_filename);
 
   if (G_OBJECT_CLASS (udisks_config_manager_parent_class))
     G_OBJECT_CLASS (udisks_config_manager_parent_class)->constructed (object);
-- 
2.20.1

